src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -80,18 +78,188 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
+                // results to use next time around.
+                var (tokens, isCSharpFinalized) = await GetSemanticTokensAsync(
+                    textDocumentIdentifier, documentSnapshot, documentVersion, range, cancellationToken);
+                if (isCSharpFinalized && tokens is not null)
+                {
+                    _tokensCache.CacheTokens(textDocumentIdentifier.Uri, semanticVersion, range, tokens.Data.ToArray());
+                }
+
+                return tokens;
+            }
+
+            var cachedRange = cachedResult.Value.Range;
+            var cachedTokens = cachedResult.Value.Tokens;
+
+            // The cache returned to us an exact match for the range we're looking for, so we can return early.
+            if (cachedRange.Equals(range))
+            {
+                return new SemanticTokens { Data = cachedTokens.ToImmutableArray() };
             }
 
-            var tokens = await GetSemanticTokensAsync(
-                textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion, range, cancellationToken);
-            return tokens;
+            // The cache returned to us only a partial match for the range. We'll need to compute the rest by
+            // sending range requests to the Razor/C#/HTML servers.
+            var filledInRange = await FillInRangeGapsAsync(
+                textDocumentIdentifier, documentSnapshot, documentVersion, semanticVersion,
+                range, cachedRange, cachedTokens, cancellationToken).ConfigureAwait(false);
+
+            return filledInRange;
+        }
+
+        private async Task<SemanticTokens?> FillInRangeGapsAsync(
+            TextDocumentIdentifier textDocumentIdentifier,
+            DocumentSnapshot documentSnapshot,
+            int documentVersion,
+            VersionStamp semanticVersion,
+            Range range,
+            Range cachedRange,
+            List<int> cachedTokens,
+            CancellationToken cancellationToken)
+        {
+            var tokenList = new List<int>();
+            var absoluteLine = 0;
+
+            // There are 3 scenarios:
+            //    1) The cached results are at the start of the range
+            //    2) The cached results are in the middle of the range
+            //    3) The cached results are at the end of the range
+            // For 1) and 3), we need to send one request to the Razor/C# language servers for the missing range's tokens.
+            // For 2), we need to send two requests (one for the missing start range, and one for the missing end range).
+
+            // 1) Before processing the cached portion of the range, we might have to compute tokens for the
+            // start of the range -> start of the cached range.
+            if (cachedRange.Start.Line != range.Start.Line)
+            {
+                var partialRange = new Range { Start = range.Start, End = cachedRange.Start };
+                var tokens = await ComputeAndCacheTokensAsync(
```

Comment:
We could potentially author both start & end requests simultaneously if we don't await immediately.

Also if we find that doing two requests leaves too much opportunity for de-synchronization (we fail on requests that are de-synchronized) we could potentially have our customer C# endpoint on the client take in these ranges and author both requests there.

I do worry that doing two separate requests leaves too much opportunity for failure where what do we do if one request succeeds and the other fails? @ryanbrandenburg please cross-check my concern here though, you'd know wayyyyy better than I

Reply:
If we do end up in a de-synchronized state, we return no tokens, so I guess you're right in that two requests would be riskier. However, I think the two-requests situation is usually quite rare. If scrolling through the file, generally only one request should be sent. I believe two requests might come into play when doing something such as navigating directly to a specific line in the file.

Good call on authoring both requests at the same time though! I'll see if I can look into making that change 👍